plugins(ubuntu-kernel): Add new plugin for Ubuntu kernels

stewarthore opened the pull request:
The new plugin standardises the kernel build process so it can be used for local development and product kernels alike. The plugin supports building an Ubuntu kernel from source, using the Ubuntu kernel build process, or from a prebuilt Ubuntu kernel debian package.

This plugin can sit alongside the existing kernel plugin. The new plugin supports creating Ubuntu kernel snaps, while the existing plugin supports creating custom kernel builds using the non-Ubuntu specific Linux kernel build process.

The plugin uses Jinja2 templates to generate the kernel build scripts. The generated scripts have been manually checked with `shellcheck -x`.

stewarthore commented:
> Hi @stewarthore, how will this integrate with the upcoming kernel and initrd plugin changes in #4302 and #5437?

This plugin implements the kernel part of KE030 Snapcraft Support for Building Kernel Snaps spec. I didn't work on this specification directly, but the rationale in the document indicates there is a need to offer a simpler kernel (and initrd) Ubuntu Core plugin. I don't believe KE030 intends to completely replace the existing kernel plugin but instead sit alongside it.

bepri commented on tests/unit/parts/plugins/test_ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,985 @@
+class BuildParameters:
+    """Parameters for the build normally provided by snapcraft."""
+
+    def __init__(
+        self,
+        base: str,
+        arch_build_on: str,
+        arch_build_for: str,
+        arch_triplet_build_for: str,
+    ):
+        self.base = base
+        self.arch_build_for = arch_build_for
+        self.arch_build_on = arch_build_on
+        self.arch_triplet_build_for = arch_triplet_build_for
```
```suggestion
   base: str
   arch_build_for: str
   arch_build_on: str
   arch_triplet_build_for: str
```

You can just use `dataclass`' default init here.

stewarthore replied:
done

bepri commented on tests/unit/parts/plugins/test_ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,985 @@
+        properties = {
+            "plugin-name": "ubuntu-kernel",
+        }
+        with pytest.raises(errors.SnapcraftError):
```
```suggestion
        with pytest.raises(errors.SnapcraftError, match="<appropriate-error-message-regex>"):
```
To help make sure we're actually testing the behavior we want to test. Same goes for the other `pytest.raises` instances

stewarthore replied:
done; I updated the exception message in the plugin as well to replace `_` with `-` so the reported field name matches the yaml field name.

bepri commented on tests/unit/parts/plugins/test_ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,985 @@
+        setup_method_fixture,
+        build_cmds_environ,
+    ):
+        """test the expected pull commands when release-name provided."""
```
Looks like this docstring is repeated a few times but the actual test is different.

stewarthore replied:
thanks, cut-paste error.

stewarthore commented on snapcraft/parts/plugins/ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,408 @@
+
+from snapcraft import errors
+
+logger = logging.getLogger(__name__)
```
Thanks, done

bepri replied:
Great, thanks. Could you change those to be `emit.debug` instead? I don't think these messages need to show at lower verbosity levels.

stewarthore commented on snapcraft/parts/plugins/ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,408 @@
+import jinja2
+import pydantic
+from craft_parts import infos, plugins
+from overrides import overrides
```
done

stewarthore commented on snapcraft/parts/plugins/ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,408 @@
+        if any(x not in _AVAILABLE_TOOLS for x in value):
+            raise errors.SnapcraftError(
+                f"unknown tool provided, available tools: {_AVAILABLE_TOOLS}"
+            )
```
agreed, thanks

stewarthore commented on tests/unit/parts/plugins/test_ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,985 @@
+        "SNAP_CONTEXT": "snap-context",
+        "SNAP": "snap",
+        "SNAP_VERSION": "1.0.0",
+    }
```
done

stewarthore commented:
@bepri Thanks for the review :beers: I've addressed your comments and replied to each but I haven't resolved the comments yet. This is my first PR with this team. Do you folks prefer the reviewer "Resolve" comments they post after they're happy with all changes, or the dev. who makes the changes to signal they've addressed the comment?

bepri commented:
I don't think we have a standard opinion from within the team, but my personal preference is to let the reviewer resolve the comments so this is perfect! Thanks.

medubelko commented on snapcraft/parts/plugins/ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,413 @@
+            kernel_abi = kernel_abi_from_version(kernel_version)
+            return kernel_version, kernel_abi
+    raise errors.SnapcraftError(
+        "Failed to identify kernel version from deb package files."
```
```suggestion
        "cannot identify kernel version from Debian packages"
```

medubelko commented on snapcraft/parts/plugins/ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,413 @@
+    """
+    rem = re.match(r"(\d+\.\d+\.\d+-\d+)\.\d+", kernel_version)
+    if not rem:
+        raise errors.SnapcraftError("Failed to parse kernel version from changelog")
```
```suggestion
        raise errors.SnapcraftError("cannot parse kernel version from changelog")
```

medubelko commented on snapcraft/parts/plugins/ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,413 @@
+        """Enforce release_name and source options are mutually exclusive."""
+        if self.ubuntu_kernel_release_name and self.source:
+            raise errors.SnapcraftError(
+                "`ubuntu-kernel-release-name` and `source` are mutually exclusive"
```
```suggestion
                "cannot use 'ubuntu-kernel-release-name' and 'source' keys at same time"
```

medubelko commented on snapcraft/parts/plugins/ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,413 @@
+            )
+        if not self.ubuntu_kernel_release_name and not self.source:
+            raise errors.SnapcraftError(
+                "must provide either `ubuntu-kernel-release-name` or `source`"
```
```suggestion
                "missing either 'ubuntu-kernel-release-name' or 'source' key"
```

medubelko commented on snapcraft/parts/plugins/ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,413 @@
+            for option in blacklist_options:
+                if getattr(self, option):
+                    raise errors.SnapcraftError(
+                        "`ubuntu-kernel-use-binary-package` and "
+                        f"`{option.replace('_', '-')}` are mutually exclusive"
```
```suggestion
                        "'ubuntu-kernel-use-binary-package' and "
                        f"'{option.replace('_', '-')}' keys are mutually exclusive"
```

medubelko commented on snapcraft/parts/plugins/ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,413 @@
+            return super().get_pull_commands()
+        if not self.release_name:
+            raise errors.SnapcraftError(
+                "must provide `ubuntu-kernel-release-name` or `source`"
```
```suggestion
                "missing either 'ubuntu-kernel-release-name' or 'source' key"
```

medubelko reviewed (commented):
@stewarthore Thanks for the PR!

I left a few suggestions to make the errors more consistent with our other tools.

upils commented on snapcraft/templates/kernel/ubuntu_kernel_get_pull_commands.sh.j2:
```diff
@@ -0,0 +1,31 @@
+# Ubuntu kernel snapcraft pull sources script
+# Generated by snapcraft ubuntu-kernel plugin.
```
```suggestion
# Rendered by snapcraft ubuntu-kernel plugin.
```

upils commented on snapcraft/templates/kernel/ubuntu_kernel_get_build_commands.sh.j2:
```diff
@@ -0,0 +1,63 @@
+# ----
+env
+
+# Cleanup build arena. Rebuilding without a `snapcraft clean` leaves state
+# behind, which slows down incremental builds significantly during development.
```
question: Can you give more details? I wonder what was left behind and why it slowed down the next build.

stewarthore replied:
removed manual cleanup code

upils commented on snapcraft/templates/kernel/ubuntu_kernel_get_build_commands.sh.j2:
```diff
@@ -0,0 +1,63 @@
+
+# Cleanup build arena. Rebuilding without a `snapcraft clean` leaves state
+# behind, which slows down incremental builds significantly during development.
+cd {{ craft_part_build_dir }}
+rm -fr {{ craft_part_build_dir }}/kernel/
+rm -fr {{ craft_part_build_dir }}/unpacked-*
+rm -fr {{ craft_part_install_dir }}/*
+
+mkdir {{ craft_part_build_dir }}/kernel
```
I understand that under `{{ craft_part_build_dir }}` the script is also creating a `{{ craft_part_build_dir }}` dir. Is that on purpose or is that a mistake?

upils commented on snapcraft/templates/kernel/ubuntu_kernel_get_pull_commands.sh.j2:
```diff
@@ -0,0 +1,31 @@
+# Ubuntu kernel snapcraft pull sources script
```
I am not sure to see the value of using jinja templates here. Is that to ease debugging when looking at the rendered scripts?
Otherwise could it be simpler to define env variables for the given inputs (ubuntu_kernel_use_binary_package, release_name, etc.) and keep these scripts "pure shell" scripts? It might make them easier to read and could be automatically checked by shellcheck.

lengau commented on snapcraft/parts/plugins/ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,413 @@
+    ) -> Self:
+        """Enforce binary package and source-only options are exclusive."""
+        if self.ubuntu_kernel_use_binary_package:
+            blacklist_options = [
+                "source",
+                "ubuntu_kernel_config",
+                "ubuntu_kernel_defconfig",
+                "ubuntu_kernel_image_target",
+                "ubuntu_kernel_tools",
+                "ubuntu_kernel_dkms",
+            ]
+            for option in blacklist_options:
```
```suggestion
            conflicting_options = (
                "source",
                "ubuntu_kernel_config",
                "ubuntu_kernel_defconfig",
                "ubuntu_kernel_image_target",
                "ubuntu_kernel_tools",
                "ubuntu_kernel_dkms",
            )
            for option in conflicting_options:
```

lengau commented on snapcraft/parts/plugins/ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,413 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+# pylint: disable=line-too-long,too-many-lines,attribute-defined-outside-init
```
We don't use pylint anymore so it's a no-op.

```suggestion
```

lengau commented on snapcraft/parts/plugins/ubuntu_kernel_plugin.py:
```diff
@@ -0,0 +1,413 @@
+        emit.message("Getting build environment")
+        return (
+            {}
+            if not self._part_info.is_cross_compiling
```
What's the effect of including these variables even if we're not cross-compiling?

bepri commented:
> Do we have a way to run shellcheck on generated scripts as part of the unit test framework? I had a look but couldn't see anything obvious.

No, our usual method is to instead have the expected output of the script stored in the project tree, and then shellcheck will catch that. Just read from said file to compare against it in the unit tests. [Here's an example](https://github.com/canonical/craft-cli/blob/f5129c0efebf38e93a951a3b2da97e37bef73542/tests/integration/test_completion.py#L75) of me doing it in another repo

> I am using jinja templates because I find they are easier to develop and maintain. It is just a personal preference so I am happy to go with whatever if there is a preferred option. Migrating the jinja to something else is pretty quick and easy.

We already have Jinja in the stack (used to render `*craft init` calls), so I'm not personally strongly against it. That said, most of us spend our time in Python and so an equivalent Python script that creates it would be easier for us. I'd like others' opinions on this though.

> Is it possible to return python script from `get_build_commands` and have it run? I looked into this but it was a bit messy, returning a shell script that calls into the python script.

There's only one place we do this right now and it sticks out like a really, really sore thumb. It's in the legacy code base. How much Python is it? If it's only a few lines, I'd say it's reasonable to just echo that into a file and then run it from bash.

bepri reviewed (approved):
I'm happy with this. Out of curiosity, is there any documentation planned for this plugin?

medubelko reviewed (approved):
CLI changes LGTM.

lengau commented on snapcraft/templates/kernel/ubuntu_kernel_build_kernel_debs.sh.inc.j2:
```diff
@@ -0,0 +1,98 @@
+# --- BEGIN kernel source build ---
```
This directory is the wrong place for these jinja templates. The `templates` directory is for jinja templates for the `init` command. You could probably put it directly in an `ubuntu_kernel` directory under `parts/plugins` though.

lengau reviewed (approved):
I'm pretty happy with this but with two requests:

1. Let's put the templates in a more appropriate spot. (I like this use of jinja templates for sufficiently complex plugins)
2. Perhaps we should have a spread test for this that's marked as manual/slow? @bepri your thoughts on that?

Approving this on the condition of #1 being fixed and #2 either being added or @bepri telling me nah.

stewarthore commented:
> I'm happy to land this now or wait until the tests are added, just let me know @stewarthore.

 @bepri if we could land this now to the feature branch that would be great, then I'll add the tests in a follow up PR. I will also push up PRs for documentation and to improve the fixed kernel build dependencies.

Thanks!
